Join stake - fee boundary values tests

piotr-iohk opened the pull request:
# Issue Number

#896

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- [x] I have added some integration tests for fee boundary values
- [ ] adjusted linear_fee in genesis.yaml to use also `certificate`




# Comments
**In comments:** questions/suggestions about adjusting error messaging
<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Assign the PR to a corresponding milestone
 ✓ Acknowledge any changes required to the Wiki
-->

KtorZ commented on lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/StakePools.hs:
```diff
@@ -207,6 +209,64 @@ spec = do
+            -- I'm unable to adjust the given transaction to cover the associated fee!
+            -- In order to do so, I'd have to select one or more additional inputs,
+            -- but I can't do that without increasing the size of the transaction
+            -- beyond the acceptable limit. Note that I am only missing 42 Lovelace.
```
Hmmm. That's an interesting case which could deserve some more investigation.

piotr-iohk replied:
My understanding was that the behavior is correct, cause it doesn't make sense to delegate 0 funds. In that case adjusting error message should be enough... (?)

KtorZ replied:
It's not really about delegating 0 funds. When you create the delegation certificates, it's not like taking a snapshot of your current funds and delegating them. It's simply creates a note that tells the node that any address containing your reward account (grouped address) should be delegated to a particular pool. So, all addresses, before and after the certificate submission carrying a ref to the reward account will be delegated. 

There's no problem to create a certificate and have no funds to delegate _yet_.

KtorZ commented on lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/StakePools.hs:
```diff
@@ -207,6 +209,64 @@ spec = do
+            \I cannot join if I have not enough fee to cover" $ \ctx -> do
+            (_, p:_) <- eventually $
+                unsafeRequest @[ApiStakePool] ctx listStakePoolsEp Empty
+            let fee = 42
```
**[suggestion]**

Move this constant to be a named test data.

**Why**

Avoid magic numbers: https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#prefer-named-constants-over-magic-numbers

Commit:
Adjust genesis.yaml to make sure we use `certificate` for delegating fees

Commit:
Join stake - fee boundary values tests

KtorZ commented on lib/jormungandr/test/data/jormungandr/genesis.yaml:
```diff
@@ -49,7 +49,7 @@ blockchain_configuration:
     constant: 42
     coefficient: 0
-    certificate: 0
+    certificate: 1
```
This is actually interesting. We should probably use some non-null fee for the coefficient part as well.

Commit:
Quit stake - fee boundry values tests

Commit:
revise errMsg403WrongPool

KtorZ committed:
define a 'joinStakePool' helper in the wallet engine

Move business logic out of the API layer, and makes the top-level errors clearer

KtorZ committed:
define a 'quitStakePool' helper in the wallet engine

Move business logic out of the API layer, and makes the top-level errors clearer

KtorZ committed:
use newly defined helper in the wallet engine to clean-up API handlers

KtorZ committed:
revise API error messages for Join/Quit

KtorZ committed:
allow API transactions outputs to be an empty list!

In the context of delegation, it is perfectly fine for a transaction to have no outputs and be
only about paying for fees with inputs.

KtorZ committed:
change Join/Quit statuses to 'stable'
